protostar/utils/data_transformer.py

```diff
@@ -9,35 +9,76 @@
 from typing_extensions import Literal
 
 from protostar.utils.abi import find_abi_item
+from protostar.protostar_exception import ProtostarException
+
+
+class DataTransformerException(ProtostarException):
+    pass
 
 
 PythonData = Dict[str, Any]
 CairoOrPythonData = Union[CairoData, PythonData]
 FromPythonTransformer = Callable[[PythonData], CairoData]
+ToPythonTransformer = Callable[[CairoData], PythonData]
 
 
 def from_python_transformer(
-    contract_abi: AbiType,
-    fn_name: str,
-    mode: Literal["inputs", "outputs"],
+    contract_abi: AbiType, fn_name: str, mode: Literal["inputs", "outputs"]
 ) -> FromPythonTransformer:
     fn_abi_item = find_abi_item(contract_abi, fn_name)
     structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
 
     def transform(data: PythonData) -> CairoData:
-        return structure_transformer.from_python(fn_abi_item[mode], **data)[0]
+        try:
+            return structure_transformer.from_python(fn_abi_item[mode], **data)[0]
+        except (TypeError, ValueError) as ex:
+            raise DataTransformerException(str(ex)) from ex
 
     return transform
 
 
 def from_python_events_transformer(
-    contract_abi: AbiType,
-    event_name: str,
+    contract_abi: AbiType, event_name: str
 ) -> FromPythonTransformer:
     structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
     event_abi_item = find_abi_item(contract_abi, event_name)
 
     def transform(data: PythonData) -> CairoData:
-        return structure_transformer.from_python(event_abi_item["data"], **data)[0]
+        try:
+            return structure_transformer.from_python(event_abi_item["data"], **data)[0]
+        except (TypeError, ValueError) as ex:
+            raise DataTransformerException(str(ex)) from ex
+
+    return transform
+
+
+def to_python_transformer(
+    contract_abi: AbiType, fn_name: str, mode: Literal["inputs", "outputs"]
+) -> ToPythonTransformer:
+    fn_abi_item = find_abi_item(contract_abi, fn_name)
+    structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
+
+    def transform(data: CairoData) -> PythonData:
+        try:
+            return structure_transformer.to_python(fn_abi_item[mode], data)._asdict()
+        except (TypeError, ValueError) as ex:
+            raise DataTransformerException(str(ex)) from ex
+
+    return transform
+
+
+def to_python_events_transformer(
+    contract_abi: AbiType, event_name: str
+) -> ToPythonTransformer:
+    structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
+    event_abi_item = find_abi_item(contract_abi, event_name)
```

Comment:
```suggestion
    event_abi_item = find_abi_item(contract_abi, event_name)
    structure_transformer = CairoSerializer(identifier_manager_from_abi(contract_abi))
```

to be aligned with `to_python_transformer` impl. and please do the same in `from_python_transformer`

Reply:
DataTransformer is going to raise an exception when output is invalid which will be later caught by the validator. I know using a transformer for validation isn't particularly clean but I don't think we have much options if we want to delegate it to starknet.py. Nevertheless afaik we wanted to add this functionality to the DataTransformer either way.

Reply:
OK, would it make sense to do wrapping of ValueError/TypeError exceptions into DataTransformerException in validation code instead of here? In order to make such shenanigans more use-case specific.

Reply:
I don't think we would ever want Protostar to crash on DataTransformer fail, this way we can finish cleanly.

Reply:
hmm, right. please reorder statements then and resolve this discussion :)